src/dawn/SIR/ASTNodeIterator.h

```diff
@@ -0,0 +1,308 @@
+
+  ASTVoidIterator clone() const;
+
+  ASTVoidIterator& operator++();
+  bool operator==(const ASTVoidIterator& other) const;
+  inline ASTVoidIterator& setToEnd() {
+    isEnd_ = true;
+    return *this;
+  }
+  inline Stmt& operator*() { return this->rootStmt_; }
+  inline bool isEnd() const { return isEnd_; }
+  inline bool isVoidIter() const { return true; }
+
+protected:
+  ASTVoidIterator(Stmt& root, bool isTop) : ASTNodeIterator_Stmt(root, isTop) {}
+};
+
+template <typename RootNode, bool onlyFirstLevel>
+class ASTRange : public IteratorRange<ASTNodeIterator<RootNode, onlyFirstLevel>> {
+public:
+  using Iterator = ASTNodeIterator<RootNode, onlyFirstLevel>;
+  ASTRange(RootNode& root)
+      : IteratorRange<Iterator>(std::move(Iterator::CreateInstance(root)),
+                                std::move(Iterator::CreateInstance(root).setToEnd())) {}
+  ASTRange(const Iterator singleton)
+      : IteratorRange<Iterator>(std::move(Iterator(singleton.clone())),
+                                std::move(++Iterator(singleton.clone()))) {}
+
+  ASTRange(ASTRange&&) = default;
+  ASTRange(const ASTRange&) = delete;
+  ASTRange& operator=(ASTRange&& o) { *this = std::move(o); }
+  ASTRange& operator=(const ASTRange&) = delete;
+
+  inline Stmt& operator[](int index) const { return *std::next(this->firstIt_, index); }
+};
+
+template <typename RootNode, bool onlyFirstLevel>
+ASTRange<RootNode, onlyFirstLevel> iterateASTOver(RootNode& root) {
+  return ASTRange<RootNode, onlyFirstLevel>(root);
+}
+
+} // namespace dawn
+
+namespace std {
+template <typename GeneratorType, bool onlyFirstLevel, typename Enable>
+struct iterator_traits<dawn::ASTNodeIterator<GeneratorType, onlyFirstLevel, Enable>> {
+  using difference_type =
+      typename iterator_traits<typename dawn::BlockStmt::StatementList::iterator>::difference_type;
+  using value_type = dawn::Stmt;
+  using pointer = dawn::Stmt*;
+  using reference = dawn::Stmt&;
+  using iterator_category = std::input_iterator_tag;
+};
+
+} // namespace std
+
+//===------------------------------------------------------------------------------------------===//
+//
+//
+//     Definitions of templated classes declared before in this file
+//
+//
+//===------------------------------------------------------------------------------------------===//
+
+namespace dawn {
+
+//
+//  Definition of ASTNodeIterator_Stmt
+//
+
+template <typename ASTNode, bool onlyFirstLevel>
+ASTNodeIterator_Stmt ASTNodeIterator_Stmt::clone() const {
+  ASTNodeIterator o(rootStmt_, isTop_);
+  o.childrenIterator_ = childrenIterator_;
+  o.restIterator_ = restIterator_ == nullptr ? nullptr
+                                             : std::unique_ptr<ASTNodeIterator>(new ASTNodeIterator(
+                                                   std::move(restIterator_->clone())));
+  return o;
+}
+
+template <typename ASTNode, bool onlyFirstLevel>
+ASTNodeIterator_Stmt& ASTNodeIterator_Stmt::operator++() {
```

Comment:
Should the iterator go inside VerticalRegion's ASTs?

Reply:
Or, equivalently, should `VerticalRegionDeclStmt.getChildren()` return the statements inside `VerticalRegionDeclStmt.verticalRegion_->Ast->root_`? (easier to implement and correct from a polymorphism design perspective)

Reply:
Exactly, for the moment there's no use case. Could it be needed in the future?